tls/extensions/s2n_client_key_share.c

```diff
@@ -56,30 +59,57 @@ const s2n_extension_type s2n_client_key_share_extension = {
     .if_missing = s2n_extension_noop_if_missing,
 };
 
-static int s2n_ecdhe_supported_curves_send(struct s2n_connection *conn, struct s2n_stuffer *out)
+static int s2n_generate_preferred_key_shares(struct s2n_connection *conn, struct s2n_stuffer *out)
 {
-    notnull_check(conn);
-    notnull_check(conn->config);
+    uint8_t preferred_key_shares = conn->preferred_key_shares;
+    const struct s2n_ecc_named_curve *named_curve = NULL;
+    struct s2n_ecc_evp_params *ecc_evp_params = NULL;
 
     const struct s2n_ecc_preferences *ecc_pref = NULL;
     GUARD(s2n_connection_get_ecc_preferences(conn, &ecc_pref));
     notnull_check(ecc_pref);
 
-    const struct s2n_ecc_named_curve *named_curve = NULL;
-    struct s2n_ecc_evp_params *ecc_evp_params = NULL;
+    bool empty_keyshares = preferred_key_shares & 1;
 
-    for (uint32_t i = 0; i < ecc_pref->count; i++) {
-        ecc_evp_params = &conn->secure.client_ecc_evp_params[i];
-        named_curve = ecc_pref->ecc_curves[i];
+    if (!conn->preferred_key_shares) {
+        /* Default behavior is to generate keyshares for all curves.
+        * The bitmap to generate keyshares for all curve is 111111110 (254),
+        * i.e. all bit values set except lsb which is RESERVED for empty keyshares */
+        preferred_key_shares = GENERATE_KEYSHARES_ALL_CURVES;
+    }
 
+    for (size_t i = 1; i <= ecc_pref->count; i++) {
+        ecc_evp_params = &conn->secure.client_ecc_evp_params[i-1];
+        named_curve = ecc_pref->ecc_curves[i-1];
         ecc_evp_params->negotiated_curve = named_curve;
         ecc_evp_params->evp_pkey = NULL;
-        GUARD(s2n_ecdhe_parameters_send(ecc_evp_params, out));
+        /* If lsb is set, skip keyshare generation for all curve */
+        if (empty_keyshares) {
+            GUARD(s2n_stuffer_write_uint16(out, ecc_evp_params->negotiated_curve->iana_id));
+            GUARD(s2n_stuffer_write_uint16(out, ecc_evp_params->negotiated_curve->share_size));
+            uint8_t *data = s2n_stuffer_raw_write(out, ecc_evp_params->negotiated_curve->share_size);
+            memset(data, 0, ecc_evp_params->negotiated_curve->share_size);
+        } else if (GENERATE_KEYSHARE_FOR_CURVE_SET(preferred_key_shares)) { /* If bit other than lsb is set, generate keyshare for the corresponding curve */
+            GUARD(s2n_ecdhe_parameters_send(ecc_evp_params, out));
+        } else { /* If bit not set, skip keyshare generation for the corresponding curve */
+            GUARD(s2n_stuffer_write_uint16(out, ecc_evp_params->negotiated_curve->iana_id));
+            GUARD(s2n_stuffer_write_uint16(out, ecc_evp_params->negotiated_curve->share_size));
+            GUARD(s2n_stuffer_skip_write(out, ecc_evp_params->negotiated_curve->share_size));
+        }
     }
 
     return S2N_SUCCESS;
 }
 
+static int s2n_ecdhe_supported_curves_send(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    notnull_check(conn);
+    notnull_check(conn->config);
```

Comment:
These checks should probably be in the `s2n_generate_preferred_key_shares` function instead as `s2n_ecdhe_supported_curves_send` doesn't actual dereference any of those pointers.

Reply:
Fair enough, this change is a left-over from my earlier consolidated PR for Client HRR. For this change it makes sense to move them into `s2n_generate_preferred_key_shares`.